### PR TITLE
Current status of IMN form SAGA

### PR DESCRIPTION
**Pull request by pmhalvor**
Due to recent discussion, closer comparisons to [mtl-tsa](https://github.com/pmhalvor/mtl-tsa) are being done to ensure any learning is taking place.

### CODE REVIEW

File: code/model.py [hunk 1]
```diff
@@ -937,12 +996,12 @@ def init_components(self, subtasks):
 
         # polarity had attention before linear
         components["polarity"].update({
-            "attention": torch.nn.MultiheadAttention(cnn_dim, num_heads=1).to(torch.device(self.device)), 
+            # "attention": torch.nn.MultiheadAttention(cnn_dim, num_heads=1).to(torch.device(self.device)),
```
Comment (pmhalvor) [inline code/model.py:999]: Temporarily removed polarity self-attention for debugging.

File: code/model.py [hunk 2]
```diff
@@ -1058,23 +1117,17 @@ def forward(self, batch):
...
+            # values = polarity_output.permute(2, 0, 1)
+            # polarity_output = polarity_output.permute(2, 0, 1)  # sequence, batch, embedding
+            # polarity_output, _ = self.components["polarity"]["attention"](
+            #     polarity_output,  # query, i.e. polar cnn output w/ weights
+            #     polarity_output,  # keys, i.e. (polar cnn output).T for self attention
+            #     values,  # values should include probabilities for B and I tags
+            #     need_weights=False,
+            #     # TODO: implement attention mask?
+            # )
+            # polarity_output = polarity_output.permute(1, 2, 0)  # batch, embedding, sequence
```
Comment (pmhalvor) [inline code/model.py:1130]: Currently removed.

`bi_probs` wasn't exactly what the `gold_probs` from the IMN project. 
They used a guided warm-up, where every epoch plays  `p(epochs) = 5/exp(epochs+5)` of a role in these probabilities. As epochs increase, the affect these gold labels that's allow into the attention mech. decreases "exponentially".  

`bi_probs` only took the logits of the `B` and `I` predictions form previous subtask.

File: code/model.py [hunk 3]
```diff
@@ -1099,6 +1152,24 @@ def forward(self, batch):
...
+        # PICK UP HERE
+        # Could help guide model in correct direction
+        # predicted prob of B or I in target and expression outputs
+        # bi_probs = (target_output[:,:,1:].sum(dim=-1) + expression_output[:,:,1:].sum(dim=-1))/2.
+        # bi_probs = bi_probs[:,:polarity_output.size(2)]
+        # values =  torch.mul(
+        #     polarity_output.permute(1, 0, 2), # embedding, batch, sequence
+        #     bi_probs,
+        # ).permute(2, 1, 0)  # sequence, batch, embedding
+        return None
```
Comment (pmhalvor) [inline code/model.py:1170]: This helper function meant to create values based on how you want to train you polarity self attention.

File: code/utils.py [hunk 4]
```diff
@@ -89,13 +89,25 @@ def pad(batch, pad_id=0, ignore_id=-1):
...
-  
+
+
+def flatten(item):
+    flat_item = [
+        ele
+        for row in item
+        for ele in row 
+    ]
+    return flat_item
```
Comment (pmhalvor) [inline code/utils.py:109]: Not even needed.

File: code/utils.py [hunk 5]
```diff
@@ -238,10 +252,33 @@ def ez_score(true_labels, predict_labels, num_labels):
             true, 
             pred, 
             labels=[e for e in range(1, num_labels)],
-            average='micro',
+            average='weighted',
```
Comment (pmhalvor) [inline code/utils.py:255]: Often very close to `proportional_f1()`, but with very small differences.

File: code/utils.py [hunk 6]
```diff
@@ -238,10 +252,33 @@ def ez_score(true_labels, predict_labels, num_labels):
...
             zero_division=1,  # set score to 1 when all labels and predictions are 0
         )
-    return total/true_labels.shape[0]
+    return total/(len(true_labels) + 1e-10)
```
Comment (pmhalvor) [inline code/utils.py:258]: Because for some reason, now we got `ZeroDivisionError`.

File: code/utils.py [hunk 7]
```diff
@@ -238,10 +252,33 @@ def ez_score(true_labels, predict_labels, num_labels):
...
+            if gold_label != pred_label and pred_label > 0:
+                fp += 1
+            #FN
+            if gold_label > 0 and pred_label != gold_label:
+                fn += 1
+    prec = tp / (tp + fp + 1e-6)
+    rec = tp / (tp + fn + 1e-6)
+    f1 = 2 * prec * rec / (prec + rec + 1e-6)
+    # return prec, rec, f1
+    return f1
```
Comment (pmhalvor) [inline code/utils.py:281]: Jeremy recommended against RACL and IMN metrics, since average target scope range in English data (1-2) was much smaller than that for the Norwegian data (2-30) (double check his range limits). 

This span metric treats predictions as follows:
> - True positive: predicted label is correct
> - False positive: prediction label present, but not correct value (both `gold > 0` and `gold==0` can occur here)
> - False negative: gold label should be present, but predicted value is wrong (both `pred > 0` and `pred==0` can occur here)

Comment (pmhalvor) [reply code/utils.py:281]: Our false negative isn't really a false negative. Should be:
- predicted label is 0, but true > 0
```
if pred_label == 0 and gold_label > 0:
    fn += 1
```

Comment (pmhalvor) [reply code/utils.py:281]: Why is this called span when it doesn't even take into account scope? 

Checks token-wise F1 for joint tri-label.
Slightly more strict on presently labels, with wrong tag (`B` or `I`) than a regular F1 would be.
Recall basically measures only how many wrong 0 were predicted, i.e. wrongly estimated 1s or 2s won't show up here.

File: code/utils.py [hunk 8]
```diff
@@ -264,10 +301,13 @@ def proportional_f1(true_labels, predict_labels, num_labels):
...
             zero_division=1,  # set score to 1 when all labels and predictions are 0
         )
-    return total/true_labels.shape[0]
+    return total/(len(true_labels) + 1e-10)
 
 
 def binary_f1(golds, preds, eps=1e-7):
+    """
+    Needs to read in a single batch at a time (currently expects full data_loader)
+    """
```
Comment (pmhalvor) [inline code/utils.py:310]: Delete note.
Write doc-string.

**Review by pmhalvor — COMMENTED** [review]
A lot of studies on learning rates have been done. 

Currently have been focus on range `[1e-7, 0.1]`.

Potential problems here could be scheduler turning on too quickly, and reducing too fast right away. 

Or all learning rates have been too high. A test for `lr=1e-9` is being run on Fox rn.
Tests on [mtl-tsa](https://github.com/pmhalvor/mtl-tsa) have also shown no results on our metrics for higher learning rates. Tests with smaller learning rates are underway as well.
